superset/sqllab/permalink/api.py

```diff
@@ -88,6 +88,8 @@ def post(self) -> Response:
             state = self.add_model_schema.load(request.json)
             key = CreateSqlLabPermalinkCommand(state=state).run()
             url = url_for("SqllabView.root", key=key, _external=True)
+            if "/sqllab/?key=" in url:
+                url = url.replace("/sqllab/?key=", "/sqllab/p/")
```

Comment:
Oops, my bad. Looking at the view:

```python
    @expose("/", methods=["GET", "POST"])
    @expose("/p/<string:permalink>/", methods=["GET"])
    @has_access
    @permission_name("read")
    @event_logger.log_this
    def root(self, **kwargs: Any) -> FlaskResponse:
        payload = {}
        if form_data := request.form.get("form_data"):
            with contextlib.suppress(json.JSONDecodeError):
                payload["requested_query"] = json.loads(form_data)
        return self.render_app_template(payload)
```

Looks like this is what we want:

```suggestion
            url = url_for("SqllabView.root", permalink=key, _external=True)
```

In general, we should always build internal URLs using `url_for`.

Reply:
So after testing this locally it still will not work.It displays the same query despite copying different queries.

Reply:
Can you try this:

```python
    @expose("/", methods=["GET", "POST"], endpoint="root")
    @expose("/p/<string:permalink>/", methods=["GET"], endpoint="root_p")
    @has_access
    @permission_name("read")
    @event_logger.log_this
    def root(self, **kwargs: Any) -> FlaskResponse:
        payload = {}
        if form_data := request.form.get("form_data"):
            with contextlib.suppress(json.JSONDecodeError):
                payload["requested_query"] = json.loads(form_data)
        return self.render_app_template(payload)
```

And then do:

```python
url = url_for("SqllabView.root_p", permalink=key, _external=True)
```

Reply:
Updated the pr and separated permalink_view from root and works well.